### PR TITLE
chore: remove `col` from Note

### PR DESCRIPTION
**Pull request by david-allison**
Revived from 
* https://github.com/ankidroid/Anki-Android/pull/13936/
* https://github.com/ankidroid/Anki-Android/pull/13936/commits/44b5f1b78db1e7d18ebbd257865621407f12133b

## Purpose / Description
This completes the merging of #13936

## Fixes
* Fixes #13936

## Approach
* Apply patch
* Fix compile errors
* Remove 'test' changes, included in `TestClass`

## How Has This Been Tested?
Unit tests + quick manual runthrough of note editor, card browser and reviewing

## Learning (optional, can help others)
I could churn a few less files with a context receiver. I'll introduce this next so we don't have the "introduction of a new concept within a refactor"  problem again for receivers.

## Checklist
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Review by BrayanDSO — APPROVED** [review]

**Comment by david-allison** [comment]
I'd like to block this on https://github.com/ankidroid/Anki-Android/pull/15254 as we can get a few cleaner uses of `withCol` here

File: AnkiDroid/src/main/java/com/ichi2/anki/browser/CardBrowserViewModel.kt [hunk 1]
```diff
@@ -41,6 +41,7 @@ import com.ichi2.libanki.Card
 import com.ichi2.libanki.CardId
 import com.ichi2.libanki.Consts
 import com.ichi2.libanki.DeckId
+import com.ichi2.libanki.hasTag
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/browser/CardBrowserViewModel.kt:44]: curious 🤔 - an import change with no change in other code?

**Review by mikehardy — APPROVED** [review]
seems fine - tiniest of notes about unexpected import change. might be me misunderstanding something

File: AnkiDroid/src/main/java/com/ichi2/anki/browser/CardBrowserViewModel.kt (hunk 1 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/browser/CardBrowserViewModel.kt:44]: The calling code before and after is:

```kotlin
  undoableOp(this) {
            // ...
            val wantMark = !noteIds.all { getNote(it).hasTag("marked") }
```

In this commit:

* we have added the `col` parameter to `hasTag`: `hasTag(col, "marked")`
* we have introduced the following, which means code inside `undoableOp` does not need the `col` parameter: `hasTag("marked")`:

```kotlin
context (Collection)
fun Note.hasTag(tag: String) = this.hasTag(this@Collection, tag)
```

Therefore only the import needs to change, and the calling code can stay the same, whilst still providing the `col` parameter to the method

**Review by lukstbit — APPROVED** [review]
Changes LGTM.

**Pull request merged**